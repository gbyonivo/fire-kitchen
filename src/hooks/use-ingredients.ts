import {
  setFetchingIngredients,
  setFetchingIngredientsError,
  setIngredients,
} from "@/lib/slices/cache-slice";
import { RootState } from "@/lib/store";
import {
  Ingredient,
  isSampleIngredient,
  SampleIngredient,
} from "@/types/ingredient";
import { convertIngredientList } from "@/utils/ingridient-converter";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { AxiosError } from "axios";
import { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

export function useIngredients(): {
  ingredients: Ingredient[];
  fetching: boolean;
  error: AxiosError | string | null;
  fetchIngredients: () => void;
  ingridientCache: Record<string, Ingredient>;
} {
  const {
    ingredients,
    fetchingIngredients,
    fetchingIngredientsError,
    ingridientCache,
  } = useSelector((state: RootState) => state.cache);
  const dispatch = useDispatch();

  const fetchIngredients = useCallback(() => {
    dispatch(setFetchingIngredients());
    const fetch = async () => {
      try {
        const response = await KitchenAxios.get("list.php?i=list");
        const data = (await response.data) as { meals: unknown[] };

        if (isSampleIngredient(data.meals?.[0])) {
          dispatch(
            setIngredients(
              convertIngredientList(data.meals as unknown as SampleIngredient[])
            )
          );
        } else {
          dispatch(setFetchingIngredientsError("Invalid data"));
        }
      } catch (e) {
        dispatch(setFetchingIngredientsError(e as AxiosError));
      }
    };
    fetch();
  }, [dispatch]);

  useEffect(() => {
    if (ingredients.length > 0 || fetchingIngredients) return;
    fetchIngredients();
  }, [fetchIngredients, ingredients, fetchingIngredients]);

  return {
    ingredients,
    fetchIngredients,
    fetching: fetchingIngredients,
    error: fetchingIngredientsError,
    ingridientCache,
  };
}
