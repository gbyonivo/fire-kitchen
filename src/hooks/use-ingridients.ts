import {
  setFetchingIngridients,
  setFetchingIngridientsError,
  setIngridients,
} from "@/lib/slices/cache-slice";
import { RootState } from "@/lib/store";
import {
  Ingredient,
  isSampleIngredient,
  SampleIngredient,
} from "@/types/ingredient";
import { convertIngredientList } from "@/utils/ingridient-converter";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { AxiosError } from "axios";
import { useCallback, useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";

export function useIngridients(): {
  ingridients: Ingredient[];
  fetching: boolean;
  error: AxiosError | string | null;
  fetchIngridients: () => void;
  ingridientCache: Record<string, Ingredient>;
} {
  const {
    ingridients,
    fetchingIngridients,
    fetchingIngridientsError,
    ingridientCache,
  } = useSelector((state: RootState) => state.cache);
  const dispatch = useDispatch();

  const fetchIngridients = useCallback(() => {
    console.log("fetching ingridients 1");
    dispatch(setFetchingIngridients());
    const fetch = async () => {
      console.log("fetching ingridients 2");
      try {
        const response = await KitchenAxios.get("list.php?i=list");
        const data = (await response.data) as { meals: unknown[] };

        if (isSampleIngredient(data.meals?.[0])) {
          dispatch(
            setIngridients(
              convertIngredientList(data.meals as unknown as SampleIngredient[])
            )
          );
        } else {
          dispatch(setFetchingIngridientsError("Invalid data"));
        }
      } catch (e) {
        dispatch(setFetchingIngridientsError(e as AxiosError));
      }
    };
    fetch();
  }, [dispatch]);

  useEffect(() => {
    if (ingridients.length > 0 || fetchingIngridients) return;
    console.log("fetching ingridients");
    fetchIngridients();
  }, [fetchIngridients, ingridients, fetchingIngridients]);

  return {
    ingridients,
    fetchIngridients,
    fetching: fetchingIngridients,
    error: fetchingIngridientsError,
    ingridientCache,
  };
}
