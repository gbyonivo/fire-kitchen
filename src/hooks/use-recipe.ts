import { RootState } from "@/lib/store";
import { useCallback, useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { AxiosError } from "axios";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { convertRecipe } from "@/utils/recipe-converter";
import { updateRecipeCache } from "@/lib/slices/cache-slice";

export function useRecipe({ recipeId }: { recipeId: string }) {
  const dispatch = useDispatch();
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState<AxiosError | null>(null);

  const recipeCache = useSelector(
    (state: RootState) => state.cache.recipeCache
  );
  const recipe = recipeCache?.[recipeId];

  const fetchRecipe = useCallback(() => {
    const getRecipe = async () => {
      if (recipe) return;
      try {
        setLoading(true);
        const response = await KitchenAxios.get(`/lookup.php`, {
          params: {
            i: recipeId,
          },
        });
        const recipe = convertRecipe(response.data.meals[0]);
        if (recipe) {
          dispatch(updateRecipeCache(recipe));
        } else {
          setError(new AxiosError("Recipe not found"));
        }
        setLoading(false);
      } catch (error) {
        setLoading(false);
        setError(error as AxiosError);
      }
    };
    getRecipe();
  }, [dispatch, recipeId, recipe]);

  useEffect(() => {
    if (recipe?.id || loading || error) return;
    fetchRecipe();
  }, [fetchRecipe, recipe?.id, loading, error]);

  return {
    recipe,
    fetchRecipe,
    loading,
    error,
  };
}
