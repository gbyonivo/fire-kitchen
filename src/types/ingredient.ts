export interface SampleIngredient {
  idIngredient?: string;
  strIngredient?: string;
  strDescription?: string;
  strType?: string;
}

export function isSampleIngredient(
  ingredient: unknown
): ingredient is SampleIngredient {
  return (
    typeof ingredient === "object" &&
    ingredient !== null &&
    "idIngredient" in ingredient &&
    "strIngredient" in ingredient
  );
}

export interface Ingredient {
  id: string;
  ingredient: string;
  description?: string;
  type?: string;
}
