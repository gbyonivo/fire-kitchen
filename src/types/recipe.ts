export interface RecipeIngredient {
  name: string;
  measure: string;
}

export interface Recipe {
  id: string;
  name: string;
  category: string;
  area: string;
  instructions: string;
  thumbnail: string;
  tags: string[];
  youtube: string;
  ingredients: RecipeIngredient[];
}
