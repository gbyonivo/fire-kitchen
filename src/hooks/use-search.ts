import { searchSuccess, setSearching } from "@/lib/slices/search-slice";
import { Ingredient } from "@/types/ingredient";
import { KitchenAxios } from "@/utils/kitchen-axios";
import { convertRecipe } from "@/utils/recipe-converter";
import { useCallback } from "react";
import { useDispatch } from "react-redux";

export function useSearch() {
  const dispatch = useDispatch();

  const searchRecipe = useCallback(
    async (ingredients: Ingredient[], search: string) => {
      try {
        dispatch(setSearching());
        const searches = [];
        if (ingredients.length > 0) {
          searches.push(
            KitchenAxios.get("filter.php", {
              params: {
                i: ingredients
                  .map((ingredient) => ingredient.ingredient)
                  .join(","),
              },
            })
          );
        }
        if (search.length > 2) {
          searches.push(
            KitchenAxios.get("/search.php", {
              params: {
                s: search,
              },
            })
          );
        }

        const responses = await Promise.all(searches);
        const recipes = responses.flatMap(
          (response) => response.data.meals || []
        );
        const uniqueRecipes = recipes.reduce((acc, recipe) => {
          if (!recipe?.idMeal) {
            return acc;
          }
          return {
            ...acc,
            [recipe.idMeal]: convertRecipe(recipe),
          };
        }, {});

        console.log(recipes.length, Object.values(uniqueRecipes));

        dispatch(searchSuccess(Object.values(uniqueRecipes)));
      } catch (error) {
        console.error(error);
      }
    },
    [dispatch]
  );

  return { searchRecipe };
}
