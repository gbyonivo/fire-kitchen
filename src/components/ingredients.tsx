import { Recipe } from "@/types/recipe";
import { SidePanel } from "./common/side-panel";
import { useState } from "react";
import { Ingredient } from "@/types/ingredient";
import { useIngridients } from "@/hooks/use-ingredients";
import clsx from "clsx";

interface IngredientsProps {
  recipe: Recipe;
  className?: string;
}

export function Ingredients({ recipe, className }: IngredientsProps) {
  const { ingridientCache } = useIngridients();
  const [selectedIngredient, setSelectedIngredient] =
    useState<Ingredient | null>(null);
  const handleClose = () => {
    setSelectedIngredient(null);
  };

  return (
    <div className={`flex flex-col gap-4 ${className}`}>
      <p className="text-sm font-bold text-gray-500">Ingredients</p>
      <div className="grid grid-cols-2 gap-2">
        {recipe.ingredients.map((ingredient) => (
          <span
            key={`${ingredient.name}-${ingredient.measure}`}
            className={clsx(
              "text-sm text-left",
              ingridientCache[ingredient.name]
                ? "hover:underline cursor-pointer"
                : "text-gray-100/50"
            )}
            role="button"
            onClick={() => {
              setSelectedIngredient(ingridientCache[ingredient.name]);
            }}
          >
            {ingredient.measure} {ingredient.name}
          </span>
        ))}
      </div>
      <SidePanel
        title={selectedIngredient?.ingredient || ""}
        open={!!selectedIngredient}
        onClose={handleClose}
      >
        {!!selectedIngredient && (
          <>
            {selectedIngredient.description ? (
              <div className="text-sm text-gray-400 leading-8">
                {selectedIngredient.description}
              </div>
            ) : (
              <div className="text-sm">
                No description available but you can more on{" "}
                <a
                  href={`https://www.google.com/search?q=${selectedIngredient.ingredient}`}
                  target="_blank"
                  className="text-blue-500 hover:text-blue-600"
                >
                  Google
                </a>
              </div>
            )}
          </>
        )}
      </SidePanel>
    </div>
  );
}
