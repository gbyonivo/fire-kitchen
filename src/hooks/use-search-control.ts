import { useRef, useState } from "react";

import { useRouter, useSearchParams } from "next/navigation";

import { usePathname } from "next/navigation";
import { useSearch } from "./use-search";
import { createQueryString } from "@/utils/common";
import { useEffect } from "react";
import { Ingredient } from "@/types/ingredient";
import { useIngredients } from "./use-ingredients";

export const useSearchControl = () => {
  const pathname = usePathname();
  const router = useRouter();
  const searchParams = useSearchParams();
  const { ingridientCache } = useIngredients();
  const timeout = useRef(null);
  const { searchRecipe } = useSearch();
  const getIngredients = () => {
    return searchParams
      .get("ingredients")
      ?.split(",")
      .reduce((acc: Ingredient[], ingredient) => {
        if (ingridientCache?.[ingredient]) {
          return [...acc, ingridientCache?.[ingredient]];
        }
        return acc;
      }, []);
  };

  console.log(getIngredients());

  const [usedParams, setUsedParams] = useState<{
    search: string;
    ingredients: Ingredient[];
  }>(() => ({
    search: searchParams.get("search") || "",
    ingredients: getIngredients(),
  }));

  useEffect(() => {
    router.push(
      pathname +
        "?" +
        createQueryString({
          nameValuePairs: [
            { name: "search", value: usedParams.search },
            {
              name: "ingredients",
              value: usedParams.ingredients.map((i) => i.ingredient).join(","),
            },
          ],
          searchParams,
        })
    );
    searchRecipe(usedParams.ingredients, usedParams.search);
  }, [pathname, router, searchParams, searchRecipe, usedParams]);

  return {
    onSetParams: (params: { search: string; ingredients: Ingredient[] }) => {
      setUsedParams(params);
    },
    initialSearchValue: searchParams.get("search") || "",
    initialIngredients: getIngredients(),
  };
};
