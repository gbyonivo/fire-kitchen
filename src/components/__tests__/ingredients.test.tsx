import { fireEvent, render } from "@testing-library/react";
import { Ingredients } from "../ingredients";
import { useIngredients } from "@/hooks/use-ingredients";
import { Recipe } from "@/types/recipe";

jest.mock("@/hooks/use-ingredients");

const mockedRecipe: Recipe = {
  id: "1",
  name: "Test",
  ingredients: [
    {
      name: "Name Test",
      measure: "Measure Test",
    },
  ],
  category: "Test",
  area: "Test",
  instructions: "Test",
  thumbnail: "Test",
  tags: [],
  youtube: "Test",
};

// TODO: Sidepanel should show when ingredient is clicked

describe("Ingredients", () => {
  test.each([
    [
      {
        "Name Test": {
          id: "1",
          name: "Name Test",
          measure: "Measure Test",
          description: "Description Test",
        },
      },
    ],
    [{}],
  ])("%s should match snapshot", async (cache) => {
    (useIngredients as unknown as jest.Mock).mockReturnValue({
      ingredients: [],
      fetching: false,
      error: null,
      fetchIngredients: jest.fn(),
      ingridientCache: cache,
    });
    const { container } = render(<Ingredients recipe={mockedRecipe} />);
    expect(container).toMatchSnapshot();
  });

  test.each([
    [
      {
        "Name Test": {
          id: "1",
          name: "Name Test",
          measure: "Measure Test",
          description: "Description Test",
        },
      },
    ],
    [{}],
  ])("%s should match snapshot after ingredient is clicked", async (cache) => {
    (useIngredients as unknown as jest.Mock).mockReturnValue({
      ingredients: [],
      fetching: false,
      error: null,
      fetchIngredients: jest.fn(),
      ingridientCache: cache,
    });
    const { container } = render(<Ingredients recipe={mockedRecipe} />);
    const ingredient = container.querySelector(
      `[data-testid="ingredient-Name Test-Measure Test"]`
    );
    if (ingredient) {
      fireEvent.click(ingredient);
    }
    expect(container).toMatchSnapshot();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });
});
