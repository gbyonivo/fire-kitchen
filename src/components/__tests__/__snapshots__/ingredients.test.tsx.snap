// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ingredients { 'Name Test': [Object] } should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-4 undefined"
  >
    <p
      class="text-sm font-bold text-gray-500"
    >
      Ingredients
    </p>
    <div
      class="grid grid-cols-2 gap-2"
    >
      <span
        class="text-sm text-left hover:underline cursor-pointer"
        data-testid="ingredient-Name Test-Measure Test"
        role="button"
      >
        Measure Test
         
        Name Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Ingredients { 'Name Test': [Object] } should match snapshot after ingredient is clicked 1`] = `
<div
  aria-hidden="true"
>
  <div
    class="flex flex-col gap-4 undefined"
  >
    <p
      class="text-sm font-bold text-gray-500"
    >
      Ingredients
    </p>
    <div
      class="grid grid-cols-2 gap-2"
    >
      <span
        class="text-sm text-left hover:underline cursor-pointer"
        data-testid="ingredient-Name Test-Measure Test"
        role="button"
      >
        Measure Test
         
        Name Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Ingredients {} should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-4 undefined"
  >
    <p
      class="text-sm font-bold text-gray-500"
    >
      Ingredients
    </p>
    <div
      class="grid grid-cols-2 gap-2"
    >
      <span
        class="text-sm text-left text-gray-100/50"
        data-testid="ingredient-Name Test-Measure Test"
        role="button"
      >
        Measure Test
         
        Name Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Ingredients {} should match snapshot after ingredient is clicked 1`] = `
<div>
  <div
    class="flex flex-col gap-4 undefined"
  >
    <p
      class="text-sm font-bold text-gray-500"
    >
      Ingredients
    </p>
    <div
      class="grid grid-cols-2 gap-2"
    >
      <span
        class="text-sm text-left text-gray-100/50"
        data-testid="ingredient-Name Test-Measure Test"
        role="button"
      >
        Measure Test
         
        Name Test
      </span>
    </div>
  </div>
</div>
`;
